eclipse-collections/src/main/java/org/eclipse/collections/impl/multimap/AbstractSynchronizedMultimap.java

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (c) 2016 Shotaro Sano.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v. 1.0 which accompany this distribution.
+ * The Eclipse Public License is available at http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ */
+
+package org.eclipse.collections.impl.multimap;
+
+import java.util.Collection;
+
+import net.jcip.annotations.GuardedBy;
+import org.eclipse.collections.api.RichIterable;
+import org.eclipse.collections.api.bag.Bag;
+import org.eclipse.collections.api.block.function.Function;
+import org.eclipse.collections.api.block.function.Function0;
+import org.eclipse.collections.api.block.function.Function2;
+import org.eclipse.collections.api.block.predicate.Predicate2;
+import org.eclipse.collections.api.block.procedure.Procedure;
+import org.eclipse.collections.api.block.procedure.Procedure2;
+import org.eclipse.collections.api.map.MutableMap;
+import org.eclipse.collections.api.multimap.Multimap;
+import org.eclipse.collections.api.multimap.MutableMultimap;
+import org.eclipse.collections.api.set.SetIterable;
+import org.eclipse.collections.api.tuple.Pair;
+
+public abstract class AbstractSynchronizedMultimap<K, V> implements MutableMultimap<K, V>
+{
+    @GuardedBy("this.lock")
+    protected final MutableMultimap<K, V> delegate;
+    private final Object lock;
+
+    protected AbstractSynchronizedMultimap(MutableMultimap<K, V> newMultimap, Object newLock)
+    {
+        if (newMultimap == null)
+        {
+            throw new IllegalArgumentException("Cannot create a AbstractSynchronizedMultimap on a null multimap");
+        }
+
+        this.delegate = newMultimap;
+        this.lock = newLock == null ? this : newLock;
+    }
+
+    protected AbstractSynchronizedMultimap(MutableMultimap<K, V> newMultimap)
+    {
+        this(newMultimap, null);
+    }
+
+    protected Object getLock()
+    {
+        return this.lock;
+    }
+
+    protected Object writeReplace()
+    {
+        return new SynchronizedMultimapSerializationProxy<K, V>(this.getDelegate());
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        synchronized (this.lock)
```

Comment:
This is already somewhat discussed earlier in this thread (though not clearly concluded). If we were to clean-up, it's rather to get rid of getter method from relevant classes. It might be possible but it's a bit too widespread change for this PR so I think we can leave it and it can be addressed separately.